### PR TITLE
Fix TimeDistributed for multi-output layers

### PR DESCRIPTION
**Pull request by PavlosMelissinos**
This PR modifies TimeDistributed so that it works with layers that have multiple outputs.

~~Note: There are no tests that cover [the activity regularizer case](https://github.com/fchollet/keras/blob/master/keras/layers/wrappers.py#L210-L214) at all so TimeDistributed **will** crash if it's used on a layer that has multiple outputs and `layer.activity_regularizer is not None`.~~

### CODE REVIEW

**Comment by PavlosMelissinos** [comment]
Some tests fail. It's due to the nonlocal keyword [here](https://github.com/fchollet/keras/pull/7554/files#diff-bc86ebad3e30c808f12a35baddbd6196L178). I had to change that from global because it didn't work after the PR. Any ideas why?

**Comment by ahundt** [comment]
nonlocal is python 3.5 only so you need to use a workaround for python 2.7.

Also, any chance the bug from https://github.com/fchollet/keras/pull/7033 might be breaking things here?

**Comment by PavlosMelissinos** [comment]
Indeed, nonlocal is a python 3 feature but global doesn't work either for some reason...

**Comment by PavlosMelissinos** [comment]
I'm not sure about that bug. Doesn't seem like it's related to this problem but I could be wrong...

**Comment by ahundt** [comment]
I took a look at the code and I think the best fix might be a nonlocal class like I did in get_file(). https://github.com/fchollet/keras/pull/6535#discussion_r115638225

**Comment by PavlosMelissinos** [comment]
Thanks @ahundt, I followed your suggestion and it seems to work on my end. Hopefully no other issues will come up.

File: keras/layers/wrappers.py [hunk 1]
```diff
@@ -167,29 +167,34 @@ def compute_output_shape(self, input_shape):
             return (child_output_shape[0], timesteps) + child_output_shape[1:]
         elif isinstance(child_output_shape, list):
             return [(shape[0], timesteps) + shape[1:] for shape in child_output_shape]
-        raise ValueError('child_output_shape should be a tuple (single output) or a list (multiple outputs)')
+        msg = 'output shape should be a tuple (single output) or a list (multiple outputs)'
+        raise ValueError(msg)
 
     def call(self, inputs, training=None, mask=None):
+        class LPWrapper:
+            # hack equivalent to nonlocal keyword
```
Comment (ahundt) [inline keras/layers/wrappers.py]: replace the word hack with workaround

File: keras/layers/wrappers.py [hunk 2]
```diff
@@ -167,29 +167,34 @@ def compute_output_shape(self, input_shape):
...
                 output = self.layer.call(x, **kwargs)
                 if not isinstance(output, list):
                     output = [output]
                 uses_learning_phase = []
-                for outputi in output:
+                for idx, outputi in enumerate(output):
                     if hasattr(outputi, '_uses_learning_phase'):
                         uses_learning_phase.append(outputi._uses_learning_phase or
-                                                   uses_learning_phase)
+                                                   uses_learning_phase[idx])
```
Comment (ahundt) [inline keras/layers/wrappers.py]: What will happen here if you try to train two independent networks with two separate `model.fit` calls in a single process?

File: keras/layers/wrappers.py (hunk 1 above)
Comment (PavlosMelissinos) [inline keras/layers/wrappers.py]: Good call

File: keras/layers/wrappers.py (hunk 2 above)
Comment (PavlosMelissinos) [inline keras/layers/wrappers.py]: This part of the code is supposed to generate a flag for each output in a single layer (or model, depending on what the input is). It shouldn't matter whether there is one model training or many at the same time so I don't see how the behaviour would be affected by that but can you explain what you mean?

File: tests/keras/layers/wrappers_test.py [hunk 3]
```diff
@@ -109,6 +109,43 @@ def test_TimeDistributed():
...
+              epochs=1)
+
+    # test single TimeDistributed layer with multiple outputs
+    timesteps = 10
+    x = Input(shape=(timesteps, 3, 2))
+    y = wrappers.TimeDistributed(model)(x)
+    outer_model = Model(inputs=x, outputs=y)
+    # plot_model(model=outer_model, to_file='full_model.png', show_shapes=True)
+    outer_model.compile(optimizer='rmsprop', loss='mse')
+    outer_model.fit(x=np.random.random((48, timesteps, 3, 2)),
```
Comment (fchollet) [inline tests/keras/layers/wrappers_test.py]: You can use `train_on_batch` with a batch with e.g. 2 samples, to make your test run faster.

File: tests/keras/layers/wrappers_test.py [hunk 4]
```diff
@@ -109,6 +109,43 @@ def test_TimeDistributed():
...
+              y=[np.random.random((16, 3, 2)), np.random.random((16, 3, 2))],
+              batch_size=16,
+              epochs=1)
+
+    # test single TimeDistributed layer with multiple outputs
+    timesteps = 10
+    x = Input(shape=(timesteps, 3, 2))
+    y = wrappers.TimeDistributed(model)(x)
+    outer_model = Model(inputs=x, outputs=y)
+    # plot_model(model=outer_model, to_file='full_model.png', show_shapes=True)
```
Comment (fchollet) [inline tests/keras/layers/wrappers_test.py]: Remove

File: tests/keras/layers/wrappers_test.py [hunk 5]
```diff
@@ -109,6 +109,43 @@ def test_TimeDistributed():
...
+
+    # test single layer with multiple outputs
+    x = Input(shape=(3, 2))
+    x1 = Lambda(function=func,
+                output_shape=output_shape,
+                mask=mask)(x)
+
+    model = Model(inputs=x, outputs=x1)
+    model.compile(optimizer='rmsprop', loss='mse')
+    model.fit(x=np.random.random((16, 3, 2)),
```
Comment (fchollet) [inline tests/keras/layers/wrappers_test.py]: You can use `train_on_batch` with a batch with e.g. 2 samples, to make your test run faster.

File: keras/layers/wrappers.py [hunk 6]
```diff
@@ -163,30 +163,47 @@ def compute_output_shape(self, input_shape):
...
             # batch size matters, use rnn-based implementation
             def step(x, _):
-                global uses_learning_phase
                 output = self.layer.call(x, **kwargs)
-                if hasattr(output, '_uses_learning_phase'):
-                    uses_learning_phase = (output._uses_learning_phase or
-                                           uses_learning_phase)
+                if not isinstance(output, list):
+                    output = [output]
+                uses_learning_phase = []
```
Comment (fchollet) [inline keras/layers/wrappers.py]: You can simplify your code a lot by simply updating your boolean, instead of using a list of booleans.

File: keras/layers/wrappers.py [hunk 7]
```diff
@@ -201,22 +218,41 @@ def step(x, _):
...
+
+        for idx, yi in enumerate(y):
+            if LPWrapper.uses_learning_phase[idx]:
+                yi._uses_learning_phase = True
+        if len(y) == 1:
+            y = y[0]
         return y
 
+    def compute_mask(self, inputs, mask=None):
+        return self.layer.compute_mask(inputs, mask)
```
Comment (fchollet) [inline keras/layers/wrappers.py]: This will not work in the general case. It's also orthogonal to this PR.

File: tests/keras/layers/wrappers_test.py (hunk 5 above)
Comment (PavlosMelissinos) [inline tests/keras/layers/wrappers_test.py]: Will do, thanks, should I replace the existing ones as well or should I leave that for a different PR?

File: keras/layers/wrappers.py (hunk 7 above)
Comment (PavlosMelissinos) [inline keras/layers/wrappers.py]: Indeed, I noticed the bug a moment ago. :(

Multi-output layers don't work unless a custom mask is defined, so I thought the same would apply to a TD layer. It seems I was wrong. Could you tell me why that's not the case here?

I have a fix that is similar to what call does, as in the following:

```
def compute_mask(self, inputs, mask=None):
    input_shape = K.int_shape(inputs)
    inputs = K.reshape(inputs, (-1,) + input_shape[2:])
    return self.layer.compute_mask(inputs, mask)
```

but I'll have to check it out more exhaustively tomorrow to find out whether it's required for my use case.

For the time being, I'll just remove the method altogether since the PR seems to pass all the tests without it. What do you think?

Comment (PavlosMelissinos) [inline keras/layers/wrappers.py]: Also, the tests seem to pass despite the addition. Should there be a new test to cover this case? As a separate PR, of course.

File: keras/layers/wrappers.py (hunk 6 above)
Comment (PavlosMelissinos) [inline keras/layers/wrappers.py]: Do we not need a boolean for each output though? If not then yeah a list is not necessary but `_uses_learning_phase` is a tensor attribute so what should [this](https://github.com/PavlosMelissinos/keras/blob/17e7e612773555d2fa79fdcfe305b0a993ef4eb7/keras/layers/wrappers.py#L246-L248) be replaced with?

here's a code snippet of the part in question:
```
for idx, yi in enumerate(y):
    if LPWrapper.uses_learning_phase[idx]:
        yi._uses_learning_phase = True
```

In my understanding, the point of the global `uses_learning_phase` variable was to transfer the old value to the converted output. This is the one produced [on line 208 in the master branch of keras](https://github.com/fchollet/keras/blob/master/keras/layers/wrappers.py#L208), which I've modified to become
```
y = [K.reshape(yi, (-1, input_length) + output_shape[idx][2:])
for idx, yi in enumerate(y)])
```
What am I missing?

File: keras/layers/wrappers.py (hunk 1 above)
Comment (PavlosMelissinos) [inline keras/layers/wrappers.py]: Done

File: tests/keras/layers/wrappers_test.py (hunk 5 above)
Comment (PavlosMelissinos) [inline tests/keras/layers/wrappers_test.py]: Done. (I just changed these two lines you pointed out)

File: tests/keras/layers/wrappers_test.py (hunk 4 above)
Comment (PavlosMelissinos) [inline tests/keras/layers/wrappers_test.py]: Done

File: tests/keras/layers/wrappers_test.py (hunk 3 above)
Comment (PavlosMelissinos) [inline tests/keras/layers/wrappers_test.py]: Done

File: keras/layers/wrappers.py (hunk 7 above)
Comment (PavlosMelissinos) [inline keras/layers/wrappers.py]: Done

**Comment by PavlosMelissinos** [comment]
Travis has gone rogue :( Somehow shows PEP8 errors on image_ocr.py which I haven't touched in this PR and crashed during pip install tensorflow on another check.

Closing and reopening the PR to trigger a rebuild.

**Comment by PavlosMelissinos** [comment]
image_ocr.py is still failing but it also fails with the master branch so it's not related to the PR.

@ahundt, @fchollet could you elaborate on your respective points, if you wouldn't mind?

**Comment by PavlosMelissinos** [comment]
There's an issue when multi-output TD is used in conjunction with the TensorBoard callback. Tensorboard tries to merge the outputs (with Pack) but it's not possible when the outputs have different shapes so the program crashes.

Other than that, the PR seems to work.

Is anyone familiar enough with Tensorboard to direct me towards a possible solution?

**Comment by fchollet** [comment]
> Tensorboard tries to merge the outputs (with Pack)

What's the exact context?

**Comment by PavlosMelissinos** [comment]
I'm getting the following output:

```
Traceback (most recent call last):
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/common_shapes.py", line 654, in _call_cpp_shape_fn_impl
    input_tensors_as_shapes, status)
  File "$ENV_PATH/lib/python3.5/contextlib.py", line 66, in __exit__
    next(self.gen)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/errors_impl.py", line 466, in raise_exception_on_not_ok_status
    pywrap_tensorflow.TF_GetCode(status))
tensorflow.python.framework.errors_impl.InvalidArgumentError: Shapes must be equal rank, but are 5 and 4
        From merging shape 0 with other shapes. for 'time_distributed_1_out/values_1' (op: 'Pack') with input shapes: [?,10,32,32,128], [?,10,65536,81].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/op_def_library.py", line 490, in apply_op
    preferred_dtype=default_dtype)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 676, in internal_convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/ops/array_ops.py", line 937, in _autopacking_conversion_function
    return _autopacking_helper(v, inferred_dtype, name or "packed")
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/ops/array_ops.py", line 900, in _autopacking_helper
    return gen_array_ops._pack(elems_as_tensors, name=scope)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/ops/gen_array_ops.py", line 1949, in _pack
    result = _op_def_lib.apply_op("Pack", values=values, axis=axis, name=name)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/op_def_library.py", line 767, in apply_op
    op_def=op_def)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 2632, in create_op
    set_shapes_for_outputs(ret)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 1911, in set_shapes_for_outputs
    shapes = shape_func(op)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 1861, in call_with_requiring
    return call_cpp_shape_fn(op, require_shape_fn=True)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/common_shapes.py", line 595, in call_cpp_shape_fn
    require_shape_fn)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/common_shapes.py", line 659, in _call_cpp_shape_fn_impl
    raise ValueError(err.message)
ValueError: Shapes must be equal rank, but are 5 and 4
        From merging shape 0 with other shapes. for 'time_distributed_1_out/values_1' (op: 'Pack') with input shapes: [?,10,32,32,128], [?,10,65536,81].

		During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/common_shapes.py", line 654, in _call_cpp_shape_fn_impl
    input_tensors_as_shapes, status)
  File "$ENV_PATH/lib/python3.5/contextlib.py", line 66, in __exit__
    next(self.gen)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/errors_impl.py", line 466, in raise_exception_on_not_ok_status
    pywrap_tensorflow.TF_GetCode(status))
tensorflow.python.framework.errors_impl.InvalidArgumentError: Shapes must be equal rank, but are 5 and 4
        From merging shape 0 with other shapes. for 'time_distributed_1_out/packed' (op: 'Pack') with input shapes: [?,10,32,32,128], [?,10,65536,81].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/op_def_library.py", line 504, in apply_op
    values, as_ref=input_arg.is_ref).dtype.name
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 676, in internal_convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/ops/array_ops.py", line 937, in _autopacking_conversion_function
    return _autopacking_helper(v, inferred_dtype, name or "packed")
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/ops/array_ops.py", line 900, in _autopacking_helper
    return gen_array_ops._pack(elems_as_tensors, name=scope)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/ops/gen_array_ops.py", line 1949, in _pack
    result = _op_def_lib.apply_op("Pack", values=values, axis=axis, name=name)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/op_def_library.py", line 767, in apply_op
    op_def=op_def)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 2632, in create_op
    set_shapes_for_outputs(ret)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 1911, in set_shapes_for_outputs
    shapes = shape_func(op)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 1861, in call_with_requiring
    return call_cpp_shape_fn(op, require_shape_fn=True)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/common_shapes.py", line 595, in call_cpp_shape_fn
    require_shape_fn)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/common_shapes.py", line 659, in _call_cpp_shape_fn_impl
    raise ValueError(err.message)
ValueError: Shapes must be equal rank, but are 5 and 4
        From merging shape 0 with other shapes. for 'time_distributed_1_out/packed' (op: 'Pack') with input shapes: [?,10,32,32,128], [?,10,65536,81].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "train.py", line 160, in <module>
    train()
  File "train.py", line 137, in train
    initial_epoch=0)
  File "$ENV_PATH/lib/python3.5/site-packages/Keras-2.0.7-py3.5.egg/keras/legacy/interfaces.py", line 87, in wrapper
  File "$ENV_PATH/lib/python3.5/site-packages/Keras-2.0.7-py3.5.egg/keras/engine/training.py", line 1953, in fit_generator
  File "$ENV_PATH/lib/python3.5/site-packages/Keras-2.0.7-py3.5.egg/keras/callbacks.py", line 52, in set_model
  File "$ENV_PATH/lib/python3.5/site-packages/Keras-2.0.7-py3.5.egg/keras/callbacks.py", line 704, in set_model
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/summary/summary.py", line 192, in histogram
    tag=tag, values=values, name=scope)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/ops/gen_logging_ops.py", line 129, in _histogram_summary
    name=name)
  File "$ENV_PATH/lib/python3.5/site-packages/tensorflow/python/framework/op_def_library.py", line 508, in apply_op
    (input_name, err))
ValueError: Tried to convert 'values' to a tensor and failed. Error: Shapes must be equal rank, but are 5 and 4
        From merging shape 0 with other shapes. for 'time_distributed_1_out/packed' (op: 'Pack') with input shapes: [?,10,32,32,128], [?,10,65536,81].
```

This error goes away if I remove the Tensorboard callback.

I've tried to:
1. turn the final `train_on_batch` in the TimeDistributed test to a `fit` with a Tensorboard callback: Still passes
2. Turn the `fit` to a `fit_generator`: still passes

So I'm not sure which part of the code is at reproduce the issue.

**Comment by xn8812** [comment]
Any updates on this? I am having the exact issue right now. My model has multiple outputs and I need to use it within the Timedistributed layer. I am using the Theano backend. Is that OK to use this new feature? @PavlosMelissinos

**Comment by PavlosMelissinos** [comment]
Hey @xn8812!

No updates, sorry. :(

I've forked keras in order to suit my needs but it's not tested for the general case and I've moved on since so I'm not really using it anymore... Try to apply my patch on keras master (I think I'm using 1.0.7 on my machine but it's been a while) and let me know if you have any problems, I'll take a look when I can.

**Comment by PavlosMelissinos** [comment]
In retrospect, I don't think this patch is worth the complications that arise. If I had to do it again, I'd split functionality into multiple layers.

**Comment by fchollet** [comment]
Closing stale PR. Sorry this did not work out.
